kolibri/core/notifications/signals.py

```diff
@@ -0,0 +1,177 @@
+                        contentnode_id=None,
+                        quiz_id=None, reason=None):
+    notification = LearnerProgressNotification()
+    notification.id = uuid.uuid4().hex
+    notification.user_id = user_id
+    notification.classroom_id = group_id
+    notification.notification_type = notification_type
+    if contentnode_id:
+        notification.contentnode_id = contentnode_id
+    if lesson_id:
+        notification.lesson_id = lesson_id
+    if quiz_id:
+        notification.quiz_id = quiz_id
+    if reason:
+        notification.reason = reason
+    return notification
+
+
+@receiver(post_save, sender=ContentSummaryLog)
+def parse_summary_log(sender, instance, **kwargs):
+    if instance.progress < 1.0:
+        return
+    lessons = get_assignments(instance, instance)
+    notifications = []
+    for lesson, contentnode_id in lessons:
+        # Check if the notification has been previously saved:
+        if LearnerProgressNotification.objects.filter(user_id=instance.user_id,
+                                                      notification_type=NotificationType.Resource,
+                                                      lesson_id=lesson.id,
+                                                      contentnode_id=contentnode_id).exists():
+            continue
+        # Let's create an ResourceIndividualCompletion
+        notification = create_notification(NotificationType.Resource,
+                                           instance.user_id,
+                                           lesson.collection_id,
+                                           lesson_id=lesson.id,
+                                           contentnode_id=contentnode_id)
+        notifications.append(notification)
+        lesson_content_ids = [resource['content_id'] for resource in lesson.resources]
+
+        # Let's check if an LessonResourceIndividualCompletion needs to be created
+        user_completed = sender.objects.filter(user_id=instance.user_id, content_id__in=lesson_content_ids, progress=1.0).count()
+        if user_completed == len(lesson_content_ids):
+            if not LearnerProgressNotification.objects.filter(user_id=instance.user_id,
+                                                              notification_type=NotificationType.Lesson,
+                                                              lesson_id=lesson.id,
+                                                              classroom_id=lesson.collection_id).exists():
+                lesson_notification = create_notification(NotificationType.Lesson, instance.user_id,
+                                                          lesson.collection_id, lesson_id=lesson.id)
+                notifications.append(lesson_notification)
+
+    save_notifications(notifications)
+
+
+@receiver(post_save, sender=ExamLog)
+def parse_exam_log(sender, instance, **kwargs):
+    if not instance.closed:
+        return
+    user_classrooms = instance.user.memberships.all()
```

Comment:
This check may not be necessary. A learner can only be assigned a Lesson or Exam if they belong to a classroom. So if a Learner is completing a quiz, then that means they have been assigned an exam through a classroom.

Reply:
Done